engine/collection/message_hub/message_hub.go

```diff
@@ -0,0 +1,471 @@
+
+	// TODO(active-pacemaker): update metrics
+	//e.metrics.MessageSent(metrics.EngineClusterCompliance, metrics.MessageClusterBlockProposal)
+	//e.core.collectionMetrics.ClusterBlockProposed(block)
+
+	return nil
+}
+
+// processQueuedVote performs actual processing of model.Vote, as a result of successful invocation
+// sends vote to designated receiver.
+// No errors are expected during normal operations.
+func (h *MessageHub) processQueuedVote(packed *packedVote) error {
+	log := h.log.With().
+		Hex("collection_id", packed.vote.BlockID[:]).
+		Uint64("collection_view", packed.vote.View).
+		Hex("recipient_id", packed.recipientID[:]).
+		Logger()
+
+	log.Info().Msg("processing vote transmission request from hotstuff")
+
+	// send the vote the desired recipient
+	err := h.con.Unicast(packed.vote, packed.recipientID)
+	if err != nil {
+		log.Err(err).Msg("could not send vote")
+		return nil
+	}
+	// TODO(active-pacemaker): update metrics
+	//h.engineMetrics.MessageSent(metrics.EngineCompliance, metrics.MessageBlockVote)
+	log.Info().Msg("collection vote transmitted")
+
+	return nil
+}
+
+// processQueuedProposal performs actual processing of flow.Header, as a result of successful invocation
+// broadcasts block proposal to collection cluster.
+// No errors are expected during normal operations.
+func (h *MessageHub) processQueuedProposal(header *flow.Header) error {
+	// first, check that we are the proposer of the block
+	if header.ProposerID != h.me.NodeID() {
+		return fmt.Errorf("cannot broadcast proposal with non-local proposer (%x)", header.ProposerID)
+	}
+
+	// get the parent of the block
+	parent, err := h.headers.ByBlockID(header.ParentID)
+	if err != nil {
+		return fmt.Errorf("could not retrieve proposal parent: %w", err)
+	}
+
+	// fill in the fields that can't be populated by HotStuff
+	// TODO(active-pacemaker): will be not relevant after merging flow.Header change
+	header.ChainID = parent.ChainID
+	header.Height = parent.Height + 1
+
+	// retrieve the payload for the block
+	payload, err := h.payloads.ByBlockID(header.ID())
+	if err != nil {
+		return fmt.Errorf("could not retrieve payload for proposal: %w", err)
+	}
+
+	log := h.log.With().
+		Str("chain_id", header.ChainID.String()).
+		Uint64("block_height", header.Height).
+		Uint64("block_view", header.View).
+		Hex("block_id", logging.ID(header.ID())).
+		Hex("parent_id", header.ParentID[:]).
+		Hex("ref_block", payload.ReferenceBlockID[:]).
+		Int("transaction_count", payload.Collection.Len()).
+		Hex("parent_signer_indices", header.ParentVoterIndices).
+		Logger()
+
+	log.Debug().Msg("processing cluster broadcast request from hotstuff")
+
+	// TODO(active-pacemaker): replace with pub/sub?
+	h.hotstuff.SubmitProposal(header, parent.View) // non-blocking
+
+	// retrieve all collection nodes in our cluster
+	recipients, err := h.state.Final().Identities(filter.And(
+		filter.In(h.cluster),
+		filter.Not(filter.HasNodeID(h.me.NodeID())),
```

Comment:
we are instantiating this filter over and over again, once for each block, despite all inputs staying the same (`h.cluster` and `h.me.NodeID()`). **For the collectors** I would suggest to initialize this filter once in the constructor instead of adding a constant runtime cost. (Caution: for consensus, we should keep the dynamic filter, because the consensus message hub exists across epochs. In contrast, the collector message hub only exists for one epoch.)